File: backend/models/monica_import.go
package models

// MonicaEntityCounts mirrors the per-entity totals of a Monica account
// (kept separate from the monica package so models stays dependency-free).
type MonicaEntityCounts struct {
	Contacts   int `json:"contacts"`
	Activities int `json:"activities"`
	Notes      int `json:"notes"`
	Reminders  int `json:"reminders"`
	Calls      int `json:"calls"`
	Tasks      int `json:"tasks"`
	Gifts      int `json:"gifts"`
	Debts      int `json:"debts"`
}

// MonicaConnectRequest starts a Monica import by validating the instance
// URL and API token. The token is held in memory only and never persisted.
type MonicaConnectRequest struct {
	BaseURL  string `json:"base_url" validate:"required,max=500"`
	APIToken string `json:"api_token" validate:"required,max=4000"`
}

// MonicaConnectResponse reports what the Monica account contains.
type MonicaConnectResponse struct {
	SessionID             string             `json:"session_id"`
	Totals                MonicaEntityCounts `json:"totals"`
	EstimatedFetchSeconds int                `json:"estimated_fetch_seconds"`
}

// MonicaFetchRequest starts the background snapshot fetch.
type MonicaFetchRequest struct {
	SessionID string `json:"session_id" validate:"required"`
	// IncludeRelationships requires one extra API request per contact
	// (Monica has no account-wide relationships endpoint), so it is opt-in.
	IncludeRelationships bool `json:"include_relationships"`
	// IncludeExtras imports calls, tasks, gifts and debts as notes.
	IncludeExtras bool `json:"include_extras"`
	// CollapseReciprocalRelationships imports only one row per relationship
	// pair. Monica stores every relationship on both contacts, whereas a
	// Meerkat relationship is directed and already surfaces on the other
	// contact as an incoming relationship — so importing both halves shows
	// the same relationship twice on each contact. Off keeps both halves.
	CollapseReciprocalRelationships bool `json:"collapse_reciprocal_relationships"`
}

// Monica import phases, in order of progression.
const (
	MonicaPhaseConnecting            = "connecting"
	MonicaPhaseFetchingContacts      = "fetching_contacts"
	MonicaPhaseFetchingActivities    = "fetching_activities"
	MonicaPhaseFetchingNotes         = "fetching_notes"
	MonicaPhaseFetchingReminders     = "fetching_reminders"
	MonicaPhaseFetchingExtras        = "fetching_extras"
	MonicaPhaseFetchingRelationships = "fetching_relationships"
	MonicaPhaseBuildingPreview       = "building_preview"
	MonicaPhaseReady                 = "ready"
	MonicaPhaseImporting             = "importing"
	MonicaPhaseImportingPhotos       = "importing_photos"
	MonicaPhaseDone                  = "done"
	MonicaPhaseFailed                = "failed"
)

// MonicaImportStatus is the polling payload for fetch and import progress.
type MonicaImportStatus struct {
	SessionID  string              `json:"session_id"`
	Phase      string              `json:"phase"`
	PhaseDone  int                 `json:"phase_done"`
	PhaseTotal int                 `json:"phase_total"`
	Error      string              `json:"error,omitempty"`
	Result     *MonicaImportResult `json:"result,omitempty"` // set from phase "importing_photos" onwards
}

// MonicaRowRelatedCounts summarizes the related entities that would be
// imported alongside one contact.
type MonicaRowRelatedCounts struct {
	Activities    int `json:"activities"`
	Notes         int `json:"notes"`
	Reminders     int `json:"reminders"`
	Relationships int `json:"relationships"`
	ExtraNotes    int `json:"extra_notes"` // calls/tasks/gifts/debts converted to notes
}

// MonicaImportRowPreview extends the generic import preview row with
// Monica-specific context.
type MonicaImportRowPreview struct {
	ImportRowPreview
	Related  MonicaRowRelatedCounts `json:"related"`
	HasPhoto bool                   `json:"has_photo"`
}

// MonicaPreviewResponse contains the full preview of a fetched snapshot.
type MonicaPreviewResponse struct {
	SessionID      string                   `json:"session_id"`
	Rows           []MonicaImportRowPreview `json:"rows"`
	TotalRows      int                      `json:"total_rows"`
	ValidRows      int                      `json:"valid_rows"`
	DuplicateCount int                      `json:"duplicate_count"`
	ErrorCount     int                      `json:"error_count"`
	Totals         MonicaRowRelatedCounts   `json:"totals"`
}

// MonicaConfirmRequest executes the import with per-contact actions.
type MonicaConfirmRequest struct {
	SessionID string            `json:"session_id" validate:"required"`
	Actions   []RowImportAction `json:"actions" validate:"required,dive"`
}

// MonicaImportResult extends the generic import result with counts for the
// related entities and the asynchronous photo phase.
type MonicaImportResult struct {
	ImportResult
	ActivitiesCreated    int `json:"activities_created"`
	NotesCreated         int `json:"notes_created"`
	RemindersCreated     int `json:"reminders_created"`
	RelationshipsCreated int `json:"relationships_created"`
	// RelationshipsCollapsed counts reciprocal halves skipped because the
	// other direction was imported instead.
	RelationshipsCollapsed int `json:"relationships_collapsed"`
	PhotosQueued         int `json:"photos_queued"`
	PhotosSaved          int `json:"photos_saved"`
	PhotosFailed         int `json:"photos_failed"`
}

File: frontend/src/api/monicaImport.ts
import { apiFetch, API_BASE_URL, getAuthHeaders, parseErrorResponse } from './client';
import type { ImportRowPreview, ImportResult, RowImportAction } from './import';

// Per-entity totals of the connected Monica account
export interface MonicaEntityCounts {
  contacts: number;
  activities: number;
  notes: number;
  reminders: number;
  calls: number;
  tasks: number;
  gifts: number;
  debts: number;
}

// Response from connecting to a Monica instance
export interface MonicaConnectResponse {
  session_id: string;
  totals: MonicaEntityCounts;
  estimated_fetch_seconds: number;
}

// Phases of the fetch/import pipeline, reported by the status endpoint
export type MonicaPhase =
  | 'connecting'
  | 'fetching_contacts'
  | 'fetching_activities'
  | 'fetching_notes'
  | 'fetching_reminders'
  | 'fetching_extras'
  | 'fetching_relationships'
  | 'building_preview'
  | 'ready'
  | 'importing'
  | 'importing_photos'
  | 'done'
  | 'failed';

// Polling payload for fetch and import progress
export interface MonicaImportStatus {
  session_id: string;
  phase: MonicaPhase;
  phase_done: number;
  phase_total: number;
  error?: string;
  result?: MonicaImportResult;
}

// Related entities that would be imported alongside one contact
export interface MonicaRowRelatedCounts {
  activities: number;
  notes: number;
  reminders: number;
  relationships: number;
  extra_notes: number;
}

// Preview row extended with Monica-specific context
export interface MonicaImportRowPreview extends ImportRowPreview {
  related: MonicaRowRelatedCounts;
  has_photo: boolean;
}

// Full preview of a fetched Monica snapshot
export interface MonicaPreviewResponse {
  session_id: string;
  rows: MonicaImportRowPreview[];
  total_rows: number;
  valid_rows: number;
  duplicate_count: number;
  error_count: number;
  totals: MonicaRowRelatedCounts;
}

// Import result extended with related-entity and photo counts
export interface MonicaImportResult extends ImportResult {
  activities_created: number;
  notes_created: number;
  reminders_created: number;
  relationships_created: number;
  relationships_collapsed: number;
  photos_queued: number;
  photos_saved: number;
  photos_failed: number;
}

// Connect makes nine serialized requests to Monica at ~1/second, and a slow
// self-hosted instance or a single rate-limit backoff pushes that well past
// the default 30s timeout. Timing out here is worse than slow: the backend
// still completes and holds the session — with the API token — while the
// browser never learns the session id needed to use or cancel it.
const CONNECT_TIMEOUT_MS = 180000;

// Validate the Monica URL + API token and create an import session
export async function monicaConnect(
  baseUrl: string,
  apiToken: string
): Promise<MonicaConnectResponse> {
  const response = await apiFetch(
    `${API_BASE_URL}/contacts/import/monica/connect`,
    {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({ base_url: baseUrl, api_token: apiToken }),
    },
    CONNECT_TIMEOUT_MS
  );

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Start the background snapshot fetch
export async function monicaStartFetch(
  sessionId: string,
  options: {
    include_relationships: boolean;
    include_extras: boolean;
    // Monica stores each relationship on both contacts. When set, only one
    // row per pair is created; the other contact shows it as incoming.
    collapse_reciprocal_relationships: boolean;
  }
): Promise<void> {
  const response = await apiFetch(
    `${API_BASE_URL}/contacts/import/monica/fetch`,
    {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({ session_id: sessionId, ...options }),
    }
  );

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }
}

// Poll fetch/import progress
export async function monicaGetStatus(
  sessionId: string
): Promise<MonicaImportStatus> {
  const response = await apiFetch(
    `${API_BASE_URL}/contacts/import/monica/status?session_id=${encodeURIComponent(sessionId)}`
  );

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Load the contact preview once the snapshot is ready
export async function monicaGetPreview(
  sessionId: string
): Promise<MonicaPreviewResponse> {
  const response = await apiFetch(
    `${API_BASE_URL}/contacts/import/monica/preview?session_id=${encodeURIComponent(sessionId)}`
  );

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Execute the import with per-contact actions
export async function monicaConfirm(
  sessionId: string,
  actions: RowImportAction[]
): Promise<MonicaImportResult> {
  const response = await apiFetch(
    `${API_BASE_URL}/contacts/import/monica/confirm`,
    {
      method: 'POST',
      headers: getAuthHeaders(),
      body: JSON.stringify({ session_id: sessionId, actions }),
    }
  );

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }

  return response.json();
}

// Cancel a session, dropping the API token server-side
export async function monicaCancel(sessionId: string): Promise<void> {
  const response = await apiFetch(
    `${API_BASE_URL}/contacts/import/monica/session?session_id=${encodeURIComponent(sessionId)}`,
    { method: 'DELETE' }
  );

  if (!response.ok) {
    throw await parseErrorResponse(response);
  }
}
